Bug 1871713: Mark Manual Clusters  Upgradeable=False If 4.6 Cred Request Secrets Are Not Provisioned

dgoodwin opened the pull request:
There are two new credentials coming in 4.6, but if a manual mode user attempts to upgrade without them, the upgrade will fail.

This change encodes the secret names we know are coming, we ensure they exist and if not you are upgradeable=false. This condition does not affect 4.5.x upgrades, only 4.5 to 4.6.

Additionally to improve user experience, we are doing a hacky reconcile to get the status update that would clear upgradeable false when these secrets are created. This involves reconciling a fake cred request NamespacedName which we watch for and return early. In 4.7 we hope to break status updating out into it's own controller for a number of reasons, at which point this hack can be removed.

dgoodwin commented:
WIP as I just want to take a cleanup pass tomorrow but this appears to be working well.

dgoodwin commented:
/bugzilla refresh

joelddiaz commented on pkg/operator/credentialsrequest/credentialsrequest_controller.go:
```diff
@@ -86,6 +91,16 @@ func newReconciler(mgr manager.Manager, actuator actuator.Actuator, platType con
 
+// isFutureCredRequestSecret is used to identify if a given secret namespace + name is one we're expecting in a future
+func isFutureCredRequestSecret(namespace, name string) bool {
+	for _, nsn := range awsactuator.GetUpcomingCredSecrets() {
```
Should we just add GetUpcomingSecrets to the actuator interface and make this less AWS-specific?

dgoodwin replied:
Done.

joelddiaz commented on pkg/operator/credentialsrequest/status.go:
```diff
@@ -274,12 +276,7 @@ func computeStatusConditions(
-		Status: configv1.ConditionTrue,
-		Type:   configv1.OperatorUpgradeable,
-	}
+	upgradeableCondition := actuator.Upgradeable(mode)
```
love it!

dgoodwin commented on pkg/aws/actuator/actuator.go:
```diff
@@ -1198,3 +1209,77 @@ func isAWSCredentials(providerSpec *runtime.RawExtension) (bool, error) {
+				missingSecrets,
+				"https://docs.openshift.com/container-platform/4.5/installing/installing_aws/manually-creating-iam.html")
+		}
+	} else if mode == operatorv1.CloudCredentialsModeMint {
```
Ah I wrongly assumed that mode calculation included a secret check and would be a reliable way to get the effective mode.

Excellent point though, I will make make it a blanket check, if not in manual then root secret is required for upgrade.

dgoodwin commented on pkg/aws/actuator/actuator.go:
```diff
@@ -1198,3 +1209,77 @@ func isAWSCredentials(providerSpec *runtime.RawExtension) (bool, error) {
+		Status: configv1.ConditionTrue,
+		Type:   configv1.OperatorUpgradeable,
+	}
+	toRelease := "4.6"
```
That looks like what you'll get if the list is empty to me, am I mistaken somewhere?

joelddiaz replied:
Yes. I wrote this comment, and then forgot to remove this once I kept reading through the code and saw the looping over GetUpcomingCredSecrets().

dgoodwin commented:
Ok I think feedback is addressed.

/assign @joelddiaz 
/cc @akhil-rane

joelddiaz reviewed (commented):
/lgtm

dgoodwin commented:
@joelddiaz sorry I added another separate commit, during demo prep I thought it might be a better demo to show that changing the mode immediately updated the condition, so we now watch the CloudCredential config and reconcile all CRs if anything changes.

dgoodwin commented:
As for bugzilla I'm going to clone the bz this week and get that ported to 4.6.

joelddiaz reviewed (commented):
/lgtm

dgoodwin commented:
Added a fix for the secret annotator controller erroring and requeuing in the perfectly valid state of the root admin cred being removed after minting.

Also switched the docs link from 4.5 to 4.6 after discussing with @jeana-redhat , this URL won't be valid until launch.

dgoodwin commented:
Updated per our slack discussion.

joelddiaz reviewed (commented):
/lgtm

dgoodwin commented:
/retest

dgoodwin commented:
We want this to merge after 4.6.

/hold

/bugzilla refresh

dgoodwin commented:
@sdodson so with this PR, I am essentially bringing the problem reported by @cgwalters in #248 back into 4.5. It is technically a separate issue where we should be instantiating a dummy actuator. Should this block this PR?

If so:

Option 1: Implement the dummy actuator and include in this 4.5 PR as it's technically going to cause that problem. I will also fix in 4.6 but via a separate PR aligned to a bug for that specific pre-existing issue.

Option 2: Skip upgradeable=false if you're in mint mode with a removed admin cred, but keep upgradeable=false if you're in manual mode and a 4.6 secret is missing. This would imply a mint move with removed admin cred user could enter into a 4.6 upgrade that will fail.

I think (1) is our best option but will take some more time.

dgoodwin commented:
Disregard previous, dummy actuator is already in play and we are backporting this code in the AWS actuator, so it should not introduce the problem.

akhil-rane commented on pkg/aws/actuator/actuator.go:
```diff
@@ -1198,3 +1198,84 @@ func isAWSCredentials(providerSpec *runtime.RawExtension) (bool, error) {
+			upgradeableCondition.Reason = constants.ErrorDeterminingUpgradeableReason
+			upgradeableCondition.Message = fmt.Sprintf("Error determining if cluster can be upgraded to %s: %v",
+				toRelease, err)
+			return upgradeableCondition
```
Is this return necessary considering we are returning right after the if-else block?

dgoodwin replied:
Not necessary today, but I'm a little worried about things getting weird if someone comes later and adds a new check around line 1270 and doesn't realize that we kind of intended to return at line 1267. We'd proc this condition, not return, and then potentially overwrite this condition. That's not a crisis but it could be a little strange. Returning explicitly here after a problem seems better to me, not much downside other than 1 LOC.

akhil-rane commented on pkg/azure/passthrough.go:
```diff
@@ -21,11 +21,14 @@ import (
 	"fmt"
 	"reflect"
 
+	configv1 "github.com/openshift/api/config/v1"
+	operatorv1 "github.com/openshift/api/operator/v1"
 	minterv1 "github.com/openshift/cloud-credential-operator/pkg/apis/cloudcredential/v1"
 	"github.com/openshift/cloud-credential-operator/pkg/operator/credentialsrequest/actuator"
 	actuatoriface "github.com/openshift/cloud-credential-operator/pkg/operator/credentialsrequest/actuator"
 	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
 	"sigs.k8s.io/controller-runtime/pkg/client"
```
We have not separated the CCO and the other imports here

dgoodwin replied:
Fixing, thanks.

akhil-rane commented on pkg/operator/constants/constants.go:
```diff
@@ -43,6 +43,18 @@ const (
+	// be updated.
+	ErrorDeterminingUpgradeableReason = "ErrorDeterminingUpgradeable"
+
+	// MissingRootCredentialUpgradeableReason is used a cluster is in mint mode with the root credential removed.
```
```suggestion
	// MissingRootCredentialUpgradeableReason is used when a cluster is in mint mode with the root credential removed.
```

dgoodwin replied:
Fixing. Also saw another typo next line I'll get as well.

akhil-rane commented on pkg/operator/credentialsrequest/actuator/actuator.go:
```diff
@@ -32,6 +36,12 @@ type Actuator interface {
+	// Upgradeable returns a ClusterOperator Upgradeable condition to indicate whether or not this cluster can
+	// be safely upgraded to the next "minor" (4.y) Openshift release.
+	Upgradeable(operatorv1.CloudCredentialsMode) *configv1.ClusterOperatorStatusCondition
+	// GetUpcomingCredSecrets returns a slice of NamespacedNames for secrets holding credentials that we know are coming in the next OpenShift release.
+	// Used to pre-check if Upgradeable condition should be true or false for manual mode deployments of the credentials operator.
+	GetUpcomingCredSecrets() []types.NamespacedName
```
Is this function temporary? How are we planning to maintain this function for future releases? For example 4.6 -> 4.7 will have different set of upcoming secrets. Does it makes sense to have something like  `GetUpcomingCredSecretsForRelease(release string)`

dgoodwin replied:
My thought was that we do not need to maintain past releases, once we're on 4.7, there's no need to keep checking for the secrets that were introduced in 4.7 because those are now handled by their respective CredentialsRequests. We now want to transition to 4.8, so I was thinking we just need one release at a time in here, whatever is coming next. The function should be updated every release but wouldn't cause major problems if we forgot because the secrets should exist by that point.

akhil-rane reviewed (commented):
Sorry for a late review. I completely missed it. Few minor comments.

dgoodwin commented on pkg/aws/actuator/actuator.go:
```diff
@@ -1198,3 +1198,84 @@ func isAWSCredentials(providerSpec *runtime.RawExtension) (bool, error) {
+		}
+	} else {
+		// If in mint or passthrough, make sure the root cred secret exists, if not it must be restored prior to upgrade.
+		rootCred := types.NamespacedName{Namespace: rootAWSCredsSecretNamespace, Name: rootAWSCredsSecret}
+		secret := &corev1.Secret{}
+
+		err := a.Client.Get(context.Background(), rootCred, secret)
```
Good idea, tried just now but it looks like no, that function is wrapping the error in a way that I believe will break the IsNotFound logic below. Probably not worth the refactor.

dgoodwin commented:
Updated for @akhil-rane comments, thanks for reviewing! A couple items I think are ok in current form but let me know if you disagree after reading.

akhil-rane commented:
/lgtm

sdodson commented:
/bugzilla refresh

Patch manger, FYI, this really needs to make the current window as this will be a prerequisite for upgrades to 4.6.

sdodson commented:
/bugzilla refresh

dgoodwin commented:
/test e2e-aws

dgoodwin commented:
I swear this happened the last time I had a fix going back to 4.5 where suddenly the PR becomes corrupted and can't pass e2e, then re-opening a new one with the exact same branch passes immediately. A few more failures and I'll try this, we may need another approval.

dgoodwin commented:
/test e2e-aws

sdodson commented:
There's some 36 PRs in the queue right now for 4.5, ever time something merges it will invalidate previous successful tests and require a retest. I wouldn't watch this pot boil, just check in on Monday.

knobunc commented:
/retest
